src/BizHawk.Client.Common/ArgParser.cs

```diff
@@ -1,221 +1,235 @@
+			root.Add(/* --lua */ OptionLuaFilePath);
+			root.Add(/* --luaconsole */ OptionOpenLuaConsole);
+			root.Add(/* --mmf */ OptionMMFPath);
+			root.Add(/* --movie */ OptionMovieFilePath);
+			root.Add(/* --open-ext-tool-dll */ OptionOpenExternalTool);
+			root.Add(/* --socket-ip */ OptionSocketServerIP);
+			root.Add(/* --socket-port */ OptionSocketServerPort);
+			root.Add(/* --socket-udp */ OptionSocketServerUseUDP);
+			root.Add(/* --url-get */ OptionHTTPClientURIGET);
+			root.Add(/* --url-post */ OptionHTTPClientURIPOST);
+			root.Add(/* --userdata */ OptionUserdataUnparsedPairs);
+			root.Add(/* --version */ OptionQueryAppVersion);
+
+			Parser = new CommandLineBuilder(root)
+//				.UseVersionOption() // "cannot be combined with other arguments" which is fair enough but `--config` is crucial on NixOS
+				.UseHelp()
+//				.UseEnvironmentVariableDirective() // useless
+				.UseParseDirective()
+				.UseSuggestDirective()
+//				.RegisterWithDotnetSuggest() // intended for dotnet tools
+//				.UseTypoCorrections() // we're only using the parser, and I guess this only works with the full buy-in
+//				.UseParseErrorReporting() // we're only using the parser, and I guess this only works with the full buy-in
+//				.UseExceptionHandler() // we're only using the parser, so nothing should be throwing
+//				.CancelOnProcessTermination() // we're only using the parser, so there's not really anything to cancel
+				.Build();
+			GeneratedOptions = root.Options.Where(static o =>
+			{
+				var t = o.GetType();
+				return !t.IsGenericType || t.GetGenericTypeDefinition() != typeof(BespokeOption<>); // no there is no simpler way to do this
+			}).ToArray();
+		}
+
+		/// <return>exit code, or <see langword="null"/> if should not exit</return>
+		/// <exception cref="ArgParserException">parsing failure, or invariant broken</exception>
+		public static int? ParseArguments(out ParsedCLIFlags parsed, string[] args)
+		{
+			parsed = default;
+			var result = Parser.Parse(args);
+			if (result.Errors.Count is not 0)
+			{
+				// write all to stdout and show first in modal dialog (done in `catch` block in `Program`)
+				Console.WriteLine("failed to parse command-line arguments:");
+				foreach (var error in result.Errors) Console.WriteLine(error.Message);
+				throw new ArgParserException($"failed to parse command-line arguments: {result.Errors[0].Message}");
+			}
+			var triggeredGeneratedOption = GeneratedOptions.FirstOrDefault(o => result.FindResultFor(o) is not null);
+			if (triggeredGeneratedOption is not null)
+			{
+				// means e.g. `./EmuHawkMono.sh --help` was passed, run whatever behaviour it normally has...
+				var exitCode = result.Invoke();
+				// ...and maybe exit
+				if (exitCode is not 0
+					|| triggeredGeneratedOption.Name is "help") // `Name` may be localised meaning this won't work? I can't grok the source for `HelpOption`
 				{
-					cmdRom = arg;
+					return exitCode;
```

Comment:
This looks kinda ugly... is there no way to standardize this into a single `Invoke()` call? I'm assuming not because the way bizhawk's argument parsing system works is so fundamentally different from what the CommandLine library expects.

Maybe a first step could be to separate long-term options that must stay for the entirety of the program's lifetime and could be stored in a struct like what already exists, and options that could be passed to the mainform directly. I don't really know, this just feel weird.

I mainly dislike the fact that trying to figure out how certain arguments are used and passed through takes multiple redirection steps: Search for `--dump-name` -> find `OptionAVDumpName` -> it gets passed to new ParsedCLIFlags as `cmdDumpName` argument -> the constructor sets the field `this.cmdDumpName` -> check where that field is used.

Reply:
> I mainly dislike the fact that trying to figure out how certain arguments are used and passed through takes multiple redirection steps: Search for `--dump-name` -> find `OptionAVDumpName` -> it gets passed to new ParsedCLIFlags as `cmdDumpName` argument -> the constructor sets the field `this.cmdDumpName` -> check where that field is used.

I could rename the options to match the `ParsedCLIFlags` fields, or better yet, rename the fields to these more descriptive names. I was hesitant to do that initially because I wanted to avoid churn in `MainForm`.

Reply:
> This looks kinda ugly... is there no way to standardize this into a single `Invoke()` call? I'm assuming not because the way bizhawk's argument parsing system works is so fundamentally different from what the CommandLine library expects.

That's it exactly. The library assumes it will be your `Program` implementation, and devolve *its execution privilege* to your peasant code. And also it heavily uses `async`. See the example here: https://learn.microsoft.com/en-us/dotnet/standard/commandline/handle-termination